src/node/txorphanage.cpp

```diff
@@ -83,373 +174,487 @@ class TxOrphanageImpl final : public TxOrphanage {
     for (unsigned int i = 0; i < tx.vout.size(); i++) {
         const auto it_by_prev = m_outpoint_to_orphan_it.find(COutPoint(tx.GetHash(), i));
         if (it_by_prev != m_outpoint_to_orphan_it.end()) {
-            for (const auto& elem : it_by_prev->second) {
-                // Belt and suspenders, each orphan should always have at least 1 announcer.
-                if (!Assume(!elem->second.announcers.empty())) continue;
+            for (const auto& wtxid : it_by_prev->second) {
+                // Belt and suspenders, each entry in m_outpoint_to_orphan_it should always have at least 1 announcement.
+                auto it = index_by_wtxid.lower_bound(ByWtxidView{wtxid, MIN_PEER});
+                if (!Assume(it != index_by_wtxid.end())) continue;
 
                 // Select a random peer to assign orphan processing, reducing wasted work if the orphan is still missing
                 // inputs. However, we don't want to create an issue in which the assigned peer can purposefully stop us
                 // from processing the orphan by disconnecting.
-                auto announcer_iter = std::begin(elem->second.announcers);
-                std::advance(announcer_iter, rng.randrange(elem->second.announcers.size()));
-                auto announcer = *(announcer_iter);
-
-                // Get this source peer's work set, emplacing an empty set if it didn't exist
-                // (note: if this peer wasn't still connected, we would have removed the orphan tx already)
-                std::set<Wtxid>& orphan_work_set = m_peer_orphanage_info.try_emplace(announcer).first->second.m_work_set;
-                // Add this tx to the work set
-                orphan_work_set.insert(elem->first);
+                auto it_end = index_by_wtxid.upper_bound(ByWtxidView{wtxid, MAX_PEER});
+                const auto num_announcers{std::distance(it, it_end)};
+                if (!Assume(num_announcers > 0)) continue;
+                std::advance(it, rng.randrange(num_announcers));
+
+                if (!Assume(it->m_tx->GetWitnessHash() == wtxid)) break;
+
+                // Mark this orphan as ready to be reconsidered.
+                static constexpr auto mark_reconsidered_modifier = [](auto& ann) { ann.m_reconsider = true; };
+                m_orphans.get<ByWtxid>().modify(it, mark_reconsidered_modifier);
+
                 LogDebug(BCLog::TXPACKAGES, "added %s (wtxid=%s) to peer %d workset\n",
-                         tx.GetHash().ToString(), tx.GetWitnessHash().ToString(), announcer);
+                            it->m_tx->GetHash().ToString(), it->m_tx->GetWitnessHash().ToString(), it->m_announcer);
             }
         }
     }
 }
 
 bool TxOrphanageImpl::HaveTx(const Wtxid& wtxid) const
 {
-    return m_orphans.count(wtxid);
+    auto it_lower = m_orphans.get<ByWtxid>().lower_bound(ByWtxidView{wtxid, MIN_PEER});
+    return it_lower != m_orphans.end() && it_lower->m_tx->GetWitnessHash() == wtxid;
```

Comment:
should this be
```suggestion
    auto it_lower = m_orphans.get<ByWtxid>().lower_bound(ByWtxidView{wtxid, MIN_PEER});
    return it_lower != m_orphans.get<ByWtxid>().end() && it_lower->m_tx->GetWitnessHash() == wtxid;
```
instead, if only for consistency? There are two more places I've found where an indexed iterator is compared to a non-indexed `.end()`: `::AddAnnouncer`, `::GetTx`